# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [34]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, classification_report


# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [16]:
from google.colab import drive
drive.mount('/content/drive')
df = pd.read_csv('/content/drive/MyDrive/Submission Akhir BMLP/hasil_clustering (1).csv')
df

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


ID  Year_Birth    Income   Kidhome  Teenhome Dt_Customer   Recency  \
0      5524   -0.985345  0.235696 -0.825218 -0.929894  2012-09-04  0.307039   
1      2174   -1.235733 -0.235454  1.032559  0.906934  2014-03-08 -0.383664   
2      4141   -0.317643  0.773999 -0.825218 -0.929894  2013-08-21 -0.798086   
3      6182    1.268149 -1.022355  1.032559 -0.929894  2014-02-10 -0.798086   
4      5324    1.017761  0.241888  1.032559 -0.929894  2014-01-19  1.550305   
...     ...         ...       ...       ...       ...         ...       ...   
2235  10870   -0.150717  0.358936 -0.825218  0.906934  2013-06-13 -0.107383   
2236   4001   -1.903435  0.470432  2.890335  0.906934  2014-06-10  0.237969   
2237   7270    1.017761  0.189476 -0.825218 -0.929894  2014-01-25  1.446700   
2238   8235   -1.068807  0.679401 -0.825218  0.906934  2014-01-24 -1.419719   
2239   9405   -1.235733  0.025208  1.032559  0.906934  2012-10-15 -0.314594   

      MntWines  MntFruits  MntMeatProducts  ...  Education_Master  \
0     0.983781   1.551577         1.679702  ...               0.0   
1    -0.870479  -0.636301        -0.713225  ...               0.0   
2     0.362723   0.570804        -0.177032  ...               0.0   
3    -0.870479  -0.560857        -0.651187  ...               0.0   
4    -0.389085   0.419916        -0.216914  ...               0.0   
...        ...        ...              ...  ...               ...   
2235  1.203678   0.419916         0.066692  ...               0.0   
2236  0.303291  -0.661449        -0.606873  ...               0.0   
2237  1.795020   0.545656         0.221789  ...               0.0   
2238  0.368666   0.092992         0.208495  ...               1.0   
2239 -0.653555  -0.586005        -0.469501  ...               0.0   

      Education_PhD  Marital_Status_Alone  Marital_Status_Divorced  \
0               0.0                   0.0                      0.0   
1               0.0                   0.0                      0.0   
2               0.0                   0.0                      0.0   
3               0.0                   0.0                      0.0   
4               1.0                   0.0                      0.0   
...             ...                   ...                      ...   
2235            0.0                   0.0                      0.0   
2236            1.0                   0.0                      0.0   
2237            0.0                   0.0                      1.0   
2238            0.0                   0.0                      0.0   
2239            1.0                   0.0                      0.0   

      Marital_Status_Married  Marital_Status_Single  Marital_Status_Together  \
0                        0.0                    1.0                      0.0   
1                        0.0                    1.0                      0.0   
2                        0.0                    0.0                      1.0   
3                        0.0                    0.0                      1.0   
4                        1.0                    0.0                      0.0   
...                      ...                    ...                      ...   
2235                     1.0                    0.0                      0.0   
2236                     0.0                    0.0                      1.0   
2237                     0.0                    0.0                      0.0   
2238                     0.0                    0.0                      1.0   
2239                     1.0                    0.0                      0.0   

      Marital_Status_Widow  Marital_Status_YOLO  Cluster_Label  
0                      0.0                  0.0              2  
1                      0.0                  0.0              1  
2                      0.0                  0.0              2  
3                      0.0                  0.0              2  
4                      0.0                  0.0              2  
...                    ...             

In [17]:
# Menghapus kolom 'Dt_Customer' dari DataFrame
df.drop(columns=['Dt_Customer'], inplace=True)

# Mengecek jumlah nilai kosong (missing value) di setiap kolom
missing_values = df.isna().sum()
missing_values


ID                         0
Year_Birth                 0
Income                     0
Kidhome                    0
Teenhome                   0
Recency                    0
MntWines                   0
MntFruits                  0
MntMeatProducts            0
MntFishProducts            0
MntSweetProducts           0
MntGoldProds               0
NumDealsPurchases          0
NumWebPurchases            0
NumCatalogPurchases        0
NumStorePurchases          0
NumWebVisitsMonth          0
AcceptedCmp3               0
AcceptedCmp4               0
AcceptedCmp5               0
AcceptedCmp1               0
AcceptedCmp2               0
Complain                   0
Z_CostContact              0
Z_Revenue                  0
Response                   0
Tahun                      0
Bulan                      0
Hari                       0
Education_Basic            0
Education_Graduation       0
Education_Master           0
Education_PhD              0
Marital_Status_Alone       0
Marital_Status_Divorced    0
Marital_Status_Married     0
Marital_Status_Single      0
Marital_Status_Together    0
Marital_Status_Widow       0
Marital_Status_YOLO        0
Cluster_Label              0
dtype: int64

# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [37]:
# Memisahkan data fitur dan target
fitur = df.drop(columns='Cluster_Label')
target = df['Cluster_Label']

# Split data: 80% training dan 20% testing
data_latih, data_uji, label_latih, label_uji = train_test_split(fitur, target, test_size=0.2, random_state=42)

# Menampilkan bentuk/dimensi dari setiap set data
print(f"Jumlah data X (latih): {data_latih.shape}")
print(f"Jumlah data X (uji)  : {data_uji.shape}")
print(f"Jumlah label y (latih): {label_latih.shape}")
print(f"Jumlah label y (uji)  : {label_uji.shape}")


Jumlah data X (latih): (1792, 40)
Jumlah data X (uji)  : (448, 40)
Jumlah label y (latih): (1792,)
Jumlah label y (uji)  : (448,)


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [43]:
# Membuat instance dari Random Forest
model = RandomForestClassifier(random_state=42)

# Proses pelatihan model dengan data training
model.fit(data_latih, label_latih)


RandomForestClassifier(random_state=42)

Tulis narasi atau penjelasan algoritma yang Anda gunakan.



Dalam proyek ini, saya menggunakan algoritma *Random Forest Classifier* untuk menyelesaikan permasalahan klasifikasi. Secara sederhana, algoritma ini digunakan untuk mengelompokkan data pelanggan berdasarkan ciri-ciri yang mereka miliki. Random Forest berperan sebagai alat bantu dalam mengambil keputusan klasifikasi secara kolektif.

**Bagaimana Algoritma Ini Bekerja:**

**langkah pertama: Pendekatan Kolektif (*Ensemble Learning*):**  
Random Forest merupakan bagian dari metode ensemble, yaitu teknik yang menggabungkan hasil dari banyak model sederhana (dalam hal ini pohon keputusan) untuk mendapatkan prediksi yang lebih kuat dan stabil. Ibaratnya seperti meminta pendapat dari banyak orang untuk mengambil keputusan terbaik, bukan hanya dari satu orang saja.

**Langkah kedua: Pembuatan Pohon Keputusan yang Beragam:**

- Algoritma ini membuat banyak *decision tree* (pohon keputusan).
- Setiap pohon dilatih menggunakan data yang diambil secara acak dari dataset asli melalui teknik *bootstrap sampling* — yaitu pengambilan sampel dengan pengembalian. Sehingga, setiap pohon belajar dari subset data yang berbeda-beda.
- Saat membentuk struktur pohon, hanya sebagian fitur yang dipilih secara acak di tiap titik percabangan. Teknik ini disebut *feature randomness*, dan bertujuan agar tiap pohon memiliki pola yang unik. Dengan begitu, model menjadi lebih tangguh dan tidak mudah overfitting terhadap data pelatihan.

### **Langkah Ketiga: Proses Prediksi**

Saat model menerima data baru yang perlu diklasifikasikan, data tersebut akan dikirim ke seluruh pohon keputusan yang ada di dalam Random Forest.  
Masing-masing pohon akan memberikan hasil prediksinya secara independen.  
Untuk tugas klasifikasi seperti pada kasus ini, keputusan akhir diambil berdasarkan hasil *voting* mayoritas—yaitu kelas yang paling banyak dipilih oleh pohon-pohon tersebut akan dijadikan sebagai hasil prediksi model.



## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

Tulis hasil evaluasi algoritma yang digunakan, jika Anda menggunakan 2 algoritma, maka bandingkan hasilnya.

In [44]:
# Prediksi menggunakan data pengujian
hasil_prediksi = rf_model.predict(data_uji)

# Evaluasi kinerja model
akurasi_model = accuracy_score(label_uji, hasil_prediksi)
nilai_f1 = f1_score(label_uji, hasil_prediksi, average='weighted')  # Cocok untuk klasifikasi multi-kelas

# Tampilkan metrik evaluasi
print("Nilai Akurasi:", akurasi_model)
print("Skor F1 (Weighted):", nilai_f1)


Nilai Akurasi: 1.0
Skor F1 (Weighted): 1.0




### **Makna dari Nilai Accuracy dan F1-Score Sebesar 1.0**

**Akurasi (Accuracy) 1.0**  
Nilai akurasi sebesar 1.0 menunjukkan bahwa model berhasil mengklasifikasikan seluruh data pada set pengujian dengan benar. Tidak ada satupun prediksi yang salah, artinya model memiliki tingkat ketepatan 100%.

**F1-Score 1.0**  
F1-Score merupakan gabungan dari dua metrik penting, yaitu precision dan recall. Skor sempurna sebesar 1.0 menunjukkan bahwa model memiliki keseimbangan sempurna antara keduanya, yang berarti:

- **Precision 1.0:** Setiap instance yang diprediksi sebagai kelas positif oleh model memang benar-benar termasuk ke dalam kelas tersebut. Tidak ada kesalahan prediksi positif (False Positive = 0).
- **Recall 1.0:** Model berhasil mendeteksi semua instance yang seharusnya diklasifikasikan sebagai positif. Tidak ada data positif yang terlewat (False Negative = 0).

Dengan kata lain, model menunjukkan performa yang sangat optimal pada data uji — mampu mengidentifikasi semua kelas secara akurat tanpa melakukan kesalahan klasifikasi.


In [45]:
# Misalkan label_uji dan hasil_prediksi sudah didefinisikan
cm = confusion_matrix(label_uji, hasil_prediksi)
print("Confusion Matrix:\n", cm)

# Tampilkan classification report untuk detail lebih lanjut
print(classification_report(label_uji, hasil_prediksi))

Confusion Matrix:
 [[125   0   0]
 [  0 156   0]
 [  0   0 167]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       125
           1       1.00      1.00      1.00       156
           2       1.00      1.00      1.00       167

    accuracy                           1.00       448
   macro avg       1.00      1.00      1.00       448
weighted avg       1.00      1.00      1.00       448



### **Interpretasi Elemen-Elemen dalam Confusion Matrix**

Berikut adalah bentuk *confusion matrix* dari hasil klasifikasi model:

```
[125,   0,   0]  
[  0, 156,   0]  
[  0,   0, 167]
```

Penjelasan tiap baris:

- **Baris pertama**: Terdapat **125 data yang benar-benar berasal dari kelas 0**, dan seluruhnya diprediksi dengan tepat oleh model sebagai kelas 0. Tidak ada yang salah diklasifikasikan ke kelas lain.
- **Baris kedua**: **156 data aktual dari kelas 1** semuanya berhasil diklasifikasikan secara akurat sebagai kelas 1. Tidak terjadi salah prediksi ke kelas 0 maupun kelas 2.
- **Baris ketiga**: **167 data yang berasal dari kelas 2** juga dipetakan dengan benar oleh model. Tidak satu pun yang tersalahkan ke kelas lain.

### **Kesimpulan dari Confusion Matrix**

- **Akurasi Prediksi yang Sempurna**: Setiap prediksi model tepat sasaran. Tidak ditemukan kesalahan klasifikasi, baik itu *false positive* maupun *false negative*.
- **Nilai Diagonal**: Angka pada garis diagonal (125, 156, dan 167) menunjukkan jumlah klasifikasi yang benar untuk masing-masing kelas. Tingginya nilai-nilai ini mengindikasikan model sangat akurat.
- **Elemen di Luar Diagonal**: Semua nilai di luar diagonal bernilai nol, yang berarti **tidak ada kesalahan** dalam pengklasifikasian antar kelas.

### **Kesesuaian dengan Metode Evaluasi Lain**

Hasil ini sejalan dengan metrik evaluasi lain seperti **akurasi (accuracy)** dan **skor F1 (F1-score)** yang masing-masing bernilai **1.00**. Dengan kata lain, model mampu mengenali dan memisahkan ketiga kelas dengan sempurna saat diuji menggunakan data uji.

### **Istilah dalam Classification Report**

- **Precision**: Mengukur seberapa tepat prediksi model terhadap kelas tertentu. Artinya, dari semua prediksi yang menyatakan "positif", berapa banyak yang benar-benar sesuai.
  - **Rumus**: `Precision = True Positive / (True Positive + False Positive)`

- **Recall**: Menilai kemampuan model dalam menemukan semua data yang benar-benar positif. Semakin tinggi nilai ini, semakin baik model dalam mendeteksi data yang seharusnya terklasifikasi.
  - **Rumus**: `Recall = True Positive / (True Positive + False Negative)`

- **F1-Score**: Merupakan rata-rata harmonis antara precision dan recall, berguna untuk menyeimbangkan keduanya terutama saat data tidak seimbang.
  - **Rumus**: `F1-Score = 2 * (Precision * Recall) / (Precision + Recall)`

- **Support**: Jumlah total data aktual untuk masing-masing kelas dalam dataset uji.

### **Interpretasi Classification Report**

Laporan ini memperlihatkan evaluasi kinerja model untuk masing-masing kelas (0, 1, dan 2) berdasarkan metrik yang dijelaskan di atas.

#### **Kelas 0:**
- **Precision**: 1.00 → Semua data yang diprediksi sebagai kelas 0 benar-benar berasal dari kelas 0.
- **Recall**: 1.00 → Model berhasil mendeteksi seluruh data yang sebenarnya milik kelas 0.
- **F1-Score**: 1.00 → Kombinasi precision dan recall yang sempurna.
- **Support**: 125 data.

#### **Kelas 1 & 2:**
- Masing-masing juga menunjukkan precision, recall, dan F1-Score bernilai 1.00, mengindikasikan performa sempurna seperti halnya pada kelas 0.

#### **Akurasi Keseluruhan:**
- **Accuracy: 1.00** → Model memprediksi seluruh data uji dengan tepat tanpa satu pun kesalahan.

#### **macro avg**:
- Rata-rata sederhana dari precision, recall, dan F1-Score untuk semua kelas, tidak memperhatikan jumlah data per kelas.

#### **weighted avg**:
- Rata-rata yang mempertimbangkan jumlah data di masing-masing kelas (*support*), memberikan gambaran yang lebih adil untuk dataset yang tidak seimbang.


### **Kesimpulan**

Laporan klasifikasi ini menunjukkan bahwa model bekerja dengan **tingkat akurasi dan keandalan yang sangat tinggi** dalam membedakan data ke dalam tiga kategori yang ada. Seluruh metrik evaluasi mencapai nilai maksimal (1.00), membuktikan tidak ada kesalahan prediksi dalam data uji.




## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [48]:
# Tentukan kombinasi parameter yang akan dieksplorasi
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

#  Inisialisasi model Random Forest baru
rf_model = RandomForestClassifier(random_state=42)

# Konfigurasikan pencarian parameter terbaik menggunakan GridSearchCV
grid_search = GridSearchCV(estimator=rf_model, param_grid=param_grid,
                          cv=5, scoring='accuracy', n_jobs=-1)

# Latih model menggunakan data pelatihan untuk menemukan kombinasi parameter optimal
grid_search.fit(data_latih, label_latih)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [None, 10, 20],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 300]},
             scoring='accuracy')

In [49]:
# Menampilkan kombinasi parameter optimal dari hasil GridSearch
print("Hyperparameter terbaik:", grid_search.best_params_)

Hyperparameter terbaik: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 5, 'n_estimators': 100}


In [53]:
# Gunakan model hasil penyetelan untuk memprediksi data uji
hasil_prediksi_tuning = best_rf_model.predict(data_uji)

# Evaluasi performa model setelah penyetelan parameter
akurasi_tuning = accuracy_score(label_uji, hasil_prediksi_tuning)
f1_tuning = f1_score(label_uji, hasil_prediksi_tuning, average='weighted')

# Tampilkan hasil evaluasi
print("Akurasi (Setelah Tuning):", akurasi_tuning)
print("Skor F1 (Setelah Tuning):", f1_tuning)

Akurasi (Setelah Tuning): 1.0
Skor F1 (Setelah Tuning): 1.0


## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [56]:
# Gunakan model hasil penyetelan untuk memprediksi data uji
hasil_prediksi_tuning = best_rf_model.predict(data_uji)

# Evaluasi performa model setelah penyetelan parameter
akurasi_tuning = accuracy_score(label_uji, hasil_prediksi_tuning)
f1_tuning = f1_score(label_uji, hasil_prediksi_tuning, average='weighted')

# Tampilkan hasil evaluasi
print("Akurasi (Setelah Tuning):", akurasi_tuning)
print("Skor F1 (Setelah Tuning):", f1_tuning)

Akurasi (Setelah Tuning): 1.0
Skor F1 (Setelah Tuning): 1.0


In [57]:
# Menghasilkan dan mencetak matriks confusion dari hasil prediksi yang telah disesuaikan
cm_tuning = confusion_matrix(label_uji, hasil_prediksi_tuning)
print("Confusion Matrix (Setelah Tuning):\n", cm_tuning)

Confusion Matrix (Setelah Tuning):
 [[125   0   0]
 [  0 156   0]
 [  0   0 167]]


In [58]:
# Cetak laporan klasifikasi untuk melihat performa tiap kelas secara rinci
print(classification_report(label_uji, hasil_prediksi_tuning))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       125
           1       1.00      1.00      1.00       156
           2       1.00      1.00      1.00       167

    accuracy                           1.00       448
   macro avg       1.00      1.00      1.00       448
weighted avg       1.00      1.00      1.00       448



## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.


### **Evaluasi dan Analisis Model Random Forest**

Dalam proyek ini, model klasifikasi yang dikembangkan menggunakan algoritma **Random Forest** ditujukan untuk memprediksi kategori atau klaster pelanggan berdasarkan karakteristik tertentu.


#### **1. Evaluasi Performa Model Sebelum dan Sesudah Hyperparameter Tuning**

**Performa Awal Model:**

Sebelum dilakukan penyetelan (*tuning*) hyperparameter, model Random Forest telah menunjukkan hasil yang sangat memuaskan, yakni nilai **akurasi dan F1-Score sebesar 1.0**. Artinya, seluruh data pada set pengujian berhasil diklasifikasikan dengan tepat tanpa adanya kesalahan.

**Tuning Hyperparameter dengan Grid Search:**

Walaupun performa awal sudah maksimal, tuning tetap dilakukan dengan metode **Grid Search** untuk mengeksplorasi kombinasi parameter yang mungkin memberikan hasil lebih baik. Grid Search secara menyeluruh mencoba berbagai kombinasi parameter seperti n_estimators, max_depth, dan lainnya, untuk menemukan konfigurasi optimal.

**Performa Model Setelah Tuning:**

Menariknya, setelah proses tuning selesai, **akurasi dan F1-Score model tetap berada di angka 1.0**, sama seperti sebelum tuning dilakukan. Ini menunjukkan bahwa konfigurasi awal sudah mendekati optimal, dan tuning tambahan tidak membawa peningkatan signifikan terhadap performa model.

**Kesimpulan Perbandingan:**

Tidak terdapat perbedaan performa yang berarti antara model sebelum dan sesudah tuning, karena hasil evaluasi tetap menunjukkan nilai sempurna. Namun, proses tuning tetap berguna sebagai langkah validasi terhadap parameter yang digunakan.

#### **2. Analisis Potensi Kelemahan Model**

Dari hasil **Confusion Matrix** dan **Classification Report**, tidak ditemukan indikasi kelemahan yang mencolok:

- **Confusion Matrix:** Seluruh prediksi berada di diagonal utama, artinya semua data berhasil diklasifikasikan dengan benar ke kelas masing-masing tanpa kesalahan (False Positive maupun False Negative).
- **Classification Report:** Nilai precision, recall, dan F1-Score masing-masing kelas adalah 1.00, mempertegas performa sempurna model pada data uji.

#### **3. Rekomendasi untuk Langkah Selanjutnya**

Meski performa model terlihat sangat kuat, beberapa hal perlu diperhatikan untuk menjamin keandalannya dalam penerapan nyata:

- **Kemungkinan Overfitting:**  
  Nilai akurasi yang terlalu sempurna bisa menjadi tanda bahwa model terlalu menyesuaikan diri dengan data latih. Hal ini dapat mengurangi kemampuan model dalam mengenali pola pada data yang benar-benar baru.

  **Saran:** Terapkan teknik validasi silang (*cross-validation*) atau uji performa model pada dataset yang benar-benar berbeda untuk memverifikasi generalisasi model.

- **Tingkat Kesulitan Dataset:**  
  Bisa jadi data uji yang digunakan tergolong sederhana sehingga mudah ditebak oleh model.

  **Saran:** Pertimbangkan untuk menguji model menggunakan dataset dengan variasi yang lebih kompleks, agar bisa melihat bagaimana model menghadapi data yang lebih menantang.

### **Kesimpulan Akhir**

Model Random Forest yang dikembangkan mampu mengklasifikasikan data dengan sangat akurat, bahkan sebelum dilakukan tuning hyperparameter. Meskipun tidak ada peningkatan performa pasca-tuning, proses tersebut tetap penting sebagai bagian dari validasi model. Ke depannya, penting untuk menguji model dalam skenario yang lebih realistis untuk memastikan kemampuannya dalam menghadapi data baru dan mencegah overfitting.

